lib/vm/src/backends/singlepass.rs

```diff
@@ -34,11 +34,19 @@ pub fn backend() -> Backend {
 }
 
 pub fn set_gas(instance: &mut Instance, limit: u64) {
-    let used = GAS_LIMIT - limit;
+    let used = if limit >= GAS_LIMIT {
+        GAS_LIMIT
+    } else {
+        GAS_LIMIT - limit
+    };
     metering::set_points_used(instance, used)
 }
 
 pub fn get_gas(instance: &Instance) -> u64 {
     let used = metering::get_points_used(instance);
-    GAS_LIMIT - used
+    if used >= GAS_LIMIT {
```

Comment:
I'd only go in the special case when the default path `GAS_LIMIT - used` is invalid, i.e. `used > GAS_LIMIT`